### PR TITLE
Feature/rdss 1784

### PR DESCRIPTION
**Pull request by camilla-synbyote**

### CODE REVIEW

File: pure_adaptor/pure/base/remapper.py [hunk 1]
```diff
@@ -1,8 +1,21 @@
...
 import logging
+import re
+
+from rdsslib.taxonomy.taxonomy_client import TaxonomyGitClient, DATE_TYPE,\
+    RESOURCE_TYPE, PERSON_ROLE, ORGANISATION_TYPE, \
+    ValueNotFound, ORGANISATION_ROLE
 
 logger = logging.getLogger(__name__)
 
+TAXONOMY_SCHEMA_REPO = 'https://github.com/JiscRDSS/taxonomyschema.git'
```
Comment (michalc) [inline pure_adaptor/pure/base/remapper.py]: I don't fully understand what this is doing, but should this be pinned to a tag or a commit?

File: pure_adaptor/pure/base/remapper.py [hunk 2]
```diff
@@ -17,5 +30,81 @@ def __init__(self, config_file_path):
...
+        else:
+            rdss_name = person_role
+        mapping = self.taxonomy_client.get_by_name(
+            PERSON_ROLE, rdss_name
+        )
+        return mapping
+
+    @functions.signature({'types': ['string']})
+    def _func_person_affiliation(self, affiliation):
+        # dummy value
```
Comment (michalc) [inline pure_adaptor/pure/base/remapper.py]: Why is this needed? (Maybe worth more of a comment in the code to explain?)

File: pure_adaptor/pure/v59/research_object_mapping.txt [hunk 3]
```diff
@@ -3,28 +3,67 @@
...
-        personIdentifier: id,
-        personFirstName: name.firstName,
-        personSecondName: name.lastName,
-        personOrganisationalUnit: organisationalUnits[].name[].value
+        personUuid : person.uuid || externalPerson.uuid,
+        personGivenName : person.name[0].value || externalPerson.name[0].value,
+        personCn : name.firstName,
+        personSn : name.lastName,
+        personEntitlement : '',
+        personAffiliation : '' | person_affiliation(@),
```
Comment (michalc) [inline pure_adaptor/pure/v59/research_object_mapping.txt]: If person_affiliation is just some dummy value, actually can you just specify the dummy value here? No need for it to be a custom function? Can always add that in if/when fairly easily?

File: pure_adaptor/pure/v59/research_object_mapping.txt [hunk 4]
```diff
@@ -3,28 +3,67 @@
   objectTitle: title.value || join(`, `, title[].value),
   objectPersonRole: personAssociations[*].{
       person: {
-        personIdentifier: id,
-        personFirstName: name.firstName,
-        personSecondName: name.lastName,
-        personOrganisationalUnit: organisationalUnits[].name[].value
+        personUuid : person.uuid || externalPerson.uuid,
```
Comment (michalc) [inline pure_adaptor/pure/v59/research_object_mapping.txt]: Such a minor thing, but there is a bit of inconsistent spacing around ":" in the changes

File: pure_adaptor/pure/v59/research_object_mapping.txt [hunk 5]
```diff
@@ -3,28 +3,67 @@
...
+            personIdentifierValue : to_string(id),
+            personIdentifierType : '' | person_identifiertype(@)
+        }]
+        },
+      role: personRole[0].value | person_role(@)
     },
-  objectDescription: description.value || '',
-  objectRights: `[]`,
-  objectDate: `[]`,
+  objectDescription: description[0].value || '',
```
Comment (michalc) [inline pure_adaptor/pure/v59/research_object_mapping.txt]: just to check that the `|| ''` is definitely necessary? What would the behaviour be if it wasn't there?

File: pure_adaptor/pure/v59/research_object_mapping.txt [hunk 6]
```diff
@@ -3,28 +3,67 @@
...
   objectValue: '',
-  objectIdentifier: `[]`,
-  objectRelatedIdentifier: `[]`,
-  objectOrganisationRole: `[]`,
+  objectIdentifier:[
+        {
+        identifierValue : doi || uuid,
+        identifierType : ''
+       }],
+  objectRelatedIdentifier:  relatedProjects[*].uuid || `[]`,
```
Comment (michalc) [inline pure_adaptor/pure/v59/research_object_mapping.txt]: Also checking that `|| '[]'` is definitely necessary?

File: pure_adaptor/pure/base/remapper.py (hunk 1 above)
Comment (camilla-synbyote) [reply pure_adaptor/pure/base/remapper.py]: The purpose of the global is to supply a link to the Taxonomyschema's Git repo to the TaxonomyGitClient, which will pull down the files in the repo when it needs to make a mapping.

Personally (and am happy to listen to arguments for/against), I would like to pin this to the most recent changes in master, so that whenever there is a new addition to the taxonomy, it will also be reflected in the client. I'm hoping this will be a very temporary solution and that the taxonomy service lambda will be deployed soon so this can be deprecated.

File: pure_adaptor/pure/v59/research_object_mapping.txt (hunk 3 above)
Comment (camilla-synbyote) [reply pure_adaptor/pure/v59/research_object_mapping.txt]: That is a good point and I did try to add the dummy value there at some point, but I think JMESPath errored out. 
My original reasoning for doing it this 'hacky' way is that it would provide a template for the next developer who is not familiar and needs to quickly add a personAffiliation custom mapping. Add the Pure field instead of the `''` and then add implementation logic in the function `person_affiliation`. I guess I based this on the fact that it took some time to understand how to properly set this up in JMESpath, but am happy to try adding the dummy value directly again.

File: pure_adaptor/pure/v59/research_object_mapping.txt (hunk 5 above)
Comment (camilla-synbyote) [reply pure_adaptor/pure/v59/research_object_mapping.txt]: Paging @fmcc : where the OR clauses added to provide a layer of `failover` protection in case the field was not available in the incoming Pure message?

File: pure_adaptor/pure/v59/research_object_mapping.txt (hunk 6 above)
Comment (camilla-synbyote) [reply pure_adaptor/pure/v59/research_object_mapping.txt]: Paging @fmcc

File: pure_adaptor/pure/base/remapper.py (hunk 1 above)
Comment (michalc) [reply pure_adaptor/pure/base/remapper.py]: > so that whenever there is a new addition to the taxonomy, it will also be reflected in the client.

Hmmm... I think there are lot of scenarios where things are more difficult/can go wrong if we don't pin, and I think even temporary things need to be updated.

For example, If _not_ pinning, then actually, the adapter gets the latest version of taxonomyschema when the adapter next restarts(?). This means that if there is a change to taxonomyschema, and the adpater _isn't_ restarted, then somewhere down the line when the adapter _is_ restarted, it will get the latest version. If there is a problem, then it will be really tricky to work out that is was due to a taxonomyschema update potentially weeks/months before.

There is a small cost admittedly: if we want to make changes in the taxonomyschema and then have then in the adapter, it does require a change in both repos. However, I think to "only have changes in the Pure adapter when we make changes to the Pure adapter code" I think the cost is worth it.

File: pure_adaptor/pure/v59/research_object_mapping.txt (hunk 5 above)
Comment (fmcc) [reply pure_adaptor/pure/v59/research_object_mapping.txt]: @michalc @camilla-synbyote Yes, IIRC it was in case that the field was not available, especially when for validation we need a blank value to be of a certain type for these. The left hand expression just returns something falsey if it doesn't find anything, so it's not suitable for a default return value.

File: pure_adaptor/pure/v59/research_object_mapping.txt (hunk 6 above)
Comment (fmcc) [reply pure_adaptor/pure/v59/research_object_mapping.txt]: @michalc @camilla-synbyote Yes, it is definitely necessary for the reasons I've outlined above.

File: pure_adaptor/pure/v59/research_object_mapping.txt (hunk 3 above)
Comment (fmcc) [reply pure_adaptor/pure/v59/research_object_mapping.txt]: Sticking my nose in, but I think that the stubbed out function for a dummy value is a better approach than just adding in a default value for the reasons @camilla-synbyote has outlined.

Comment (michalc) [reply pure_adaptor/pure/v59/research_object_mapping.txt]: I think given that there are other examples for custom functions here, then I would lean to remove.

[I am also a bit in two minds about whether to keep things in the code just since they were tricky to work out originally... I usually say that clarity of the current behaviour of the code is paramount, and there are _far_ more possible things the next developer will want to achieve that _aren't_ wanting to set `personAffiliation`, so actually we are more likely to help them if we don't have unnecessary functions/layers. But it's tricky...]

File: pure_adaptor/pure/v59/research_object_mapping.txt (hunk 6 above)
Comment (michalc) [reply pure_adaptor/pure/v59/research_object_mapping.txt]: @fmcc @camilla-synbyote ok cool

File: pure_adaptor/pure/v59/research_object_mapping.txt (hunk 3 above)
Comment (michalc) [reply pure_adaptor/pure/v59/research_object_mapping.txt]: @camilla-synbyote But it's minor. If you would prefer to keep, all good.

File: pure_adaptor/pure/base/remapper.py (hunk 1 above)
Comment (camilla-synbyote) [reply pure_adaptor/pure/base/remapper.py]: Capability to check out based on tag has been added in the TaxonomyGitClient
https://github.com/JiscRDSS/rdss-shared-libraries/pull/2

File: pure_adaptor/pure/v59/research_object_mapping.txt (hunk 4 above)
Comment (camilla-synbyote) [reply pure_adaptor/pure/v59/research_object_mapping.txt]: Ok, hopefully I've caught all of them.

File: pure_adaptor/pure/base/remapper.py (hunk 1 above)
Comment (camilla-synbyote) [reply pure_adaptor/pure/base/remapper.py]: This should be addressed now.
